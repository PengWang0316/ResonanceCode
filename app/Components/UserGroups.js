import React, { Component } from 'react';

import styles from '../styles/UserGroups.module.css';
import UserGroupModal from './UserGroupModal';
import DeleteUserGroupComformModal from './DeleteUserGroupComformModal';

/** Showing the user group list. */
class UserGroups extends Component {
  state = {
    userGroup: null, groupName: '', isUpdate: false, deleteGroupName: ''
  };

  /** setting a listener to clear states when modal is closed.
    * $ will use the jQuery library from index.html page.
    * @return {null} No return.
  */
  componentDidMount() {
    $('#userGroupModal').on('hidden.bs.modal', () => {
      // this.isUpdate = false; // Reset the update indicator.
      this.setState({ userGroup: null, groupName: '', isUpdate: false }); // also reset the states for next use.
    });
  }

  /** Showing the userGroupModal when a user click a group lable.
    * $ will use jQuery libaray comes from index.html.
    * @param {object} event is an object that repersents the element a user is clicking.
    * @return {null} No return.
  */
  handleClick = ({ target }) => {
    this.setState({ userGroup: this.props.userGroups[target.parentElement.id], groupName: target.parentElement.id, isUpdate: true }, () => $('#userGroupModal').modal('toggle'));
  };

  /** Showing the adding user group modal when a user clicks add group button.
    * @return {null} No return.
  */
  handleAddGroupClick = () => {
    this.setState({ userGroup: null, groupName: '', isUpdate: false }, () => $('#userGroupModal').modal('toggle'));
  };

  /** Show a conform modal when a user clicks the delete button.
    * $ will use the jQuery libaray from index.html page.
    * @param {object} event is an object that repersents the element a user is clicking.
    * @return {null} No return.
  */
  handleRemoveGroup = ({ target }) =>
    this.setState({ deleteGroupName: target.parentElement.id }, () =>
      $('#deleteUserGroupConformModal').modal('toggle'));
  /** Rendering the jsx for the component.
    * @return {null} Return jsx for the component.
  */
  render() {
    return (
      <div>
        <div className={`card border-dark mb-3 ${styles.cardDiv}`}>
          <h5 className="card-header">Existed user groups</h5>
          <div className="card-body text-dark">
            <div className="d-flex flex-wrap">
              {this.props.userGroups &&
                 Object.keys(this.props.userGroups).map(key => (
                   <div key={key} id={key} className={`mr-3 ${styles.groupDiv}`} title={this.props.userGroups[key].map(user => user.displayName).join()}>
                     <span tabIndex="-1" role="button" onClick={this.handleClick}>{key}</span> <i role="button" className={`fa fa-times ${styles.closeBtn}`} aria-hidden="true" title="Delete this group" onClick={this.handleRemoveGroup} />
                   </div>
                 ))}
            </div>
            <div className="text-right mt-4"><button type="button" className="btn btn-primary" onClick={this.handleAddGroupClick}>Add Group</button></div>
          </div>
        </div>
        <UserGroupModal
          userGroup={this.state.userGroup}
          groupName={this.state.groupName}
          isUpdate={this.state.isUpdate}
        />
        <DeleteUserGroupComformModal groupName={this.state.deleteGroupName} />
      </div>
    );
  }
}
// const mapStateToProps = state => ({
//   user: state.user
// });
export default UserGroups;
