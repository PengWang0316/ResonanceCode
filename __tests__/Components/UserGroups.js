import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import UserGroups from '../../app/Components/UserGroups';

jest.mock('../../app/Components/UserGroupModal', () => 'UserGroupModal');
jest.mock('../../app/Components/DeleteUserGroupComformModal', () => 'DeleteUserGroupComformModal');
const mockOnFunc = jest.fn((functionName, callback) => callback());
// const mockModalFunc = jest.fn();
window.$ = jest.fn().mockReturnValue({ on: mockOnFunc });

describe('UserGroups test', () => {
  const defaultProps = {
    userGroups: { id1: [{ id: 'idA', displayName: 'displayNameA', photo: 'photoA' }] }
  };
  const getShallowComponent = (props = defaultProps) => shallow(<UserGroups {...props} />);

  test('initial states and componentDidMount', () => {
    const component = getShallowComponent();
    expect(component.state('userGroup')).toBeNull();
    expect(component.state('groupName')).toBe('');
    expect(component.state('isUpdate')).toBeFalsy();
    expect(component.state('deleteGroupName')).toBe('');
    expect(window.$).toHaveBeenLastCalledWith('#userGroupModal');
    expect(mockOnFunc).toHaveBeenCalledTimes(1);
  });

  test('UserGroups snapshot', () => expect(renderer.create(<UserGroups {...defaultProps} />).toJSON()).toMatchSnapshot());

  test('handleClick', () => {
    const mockModalFunc = jest.fn();
    window.$ = jest.fn().mockReturnValue({ on: mockOnFunc, modal: mockModalFunc });
    const component = getShallowComponent();
    component.find('span').simulate('click', { target: { parentElement: { id: 'id1' } } });
    expect(component.state('userGroup')).toEqual([{ id: 'idA', displayName: 'displayNameA', photo: 'photoA' }]);
    expect(component.state('groupName')).toBe('id1');
    expect(component.state('isUpdate')).toBeTruthy();
    expect(window.$).toHaveBeenLastCalledWith('#userGroupModal');
    expect(mockModalFunc).toHaveBeenLastCalledWith('toggle');
  });

  test('handleAddGroupClick', () => {
    const mockModalFunc = jest.fn();
    window.$ = jest.fn().mockReturnValue({ on: mockOnFunc, modal: mockModalFunc });
    const component = getShallowComponent();
    component.find('button').simulate('click');
    expect(component.state('userGroup')).toBeNull();
    expect(component.state('groupName')).toBe('');
    expect(component.state('isUpdate')).toBeFalsy();
    expect(window.$).toHaveBeenLastCalledWith('#userGroupModal');
    expect(mockModalFunc).toHaveBeenLastCalledWith('toggle');
  });

  test('handleRemoveGroup', () => {
    const mockModalFunc = jest.fn();
    window.$ = jest.fn().mockReturnValue({ on: mockOnFunc, modal: mockModalFunc });
    const component = getShallowComponent();
    component.find('i').simulate('click', { target: { parentElement: { id: 'idA' } } });
    expect(component.state('deleteGroupName')).toBe('idA');
    expect(window.$).toHaveBeenLastCalledWith('#deleteUserGroupConformModal');
    expect(mockModalFunc).toHaveBeenLastCalledWith('toggle');
  });
});
