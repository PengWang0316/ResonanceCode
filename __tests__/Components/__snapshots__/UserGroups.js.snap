// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserGroups test UserGroups snapshot 1`] = `
<div>
  <div
    className="card border-dark mb-3 cardDiv"
  >
    <h5
      className="card-header"
    >
      Existed user groups
    </h5>
    <div
      className="card-body text-dark"
    >
      <div
        className="d-flex flex-wrap"
      >
        <div
          className="mr-3 groupDiv"
          id="id1"
          title="displayNameA"
        >
          <span
            onClick={[Function]}
            role="button"
            tabIndex="-1"
          >
            id1
          </span>
           
          <i
            aria-hidden="true"
            className="fa fa-times closeBtn"
            onClick={[Function]}
            role="button"
            title="Delete this group"
          />
        </div>
      </div>
      <div
        className="text-right mt-4"
      >
        <button
          className="btn btn-primary"
          onClick={[Function]}
          type="button"
        >
          Add Group
        </button>
      </div>
    </div>
  </div>
  <UserGroupModal
    groupName=""
    isUpdate={false}
    userGroup={null}
  />
  <DeleteUserGroupComformModal
    groupName=""
  />
</div>
`;
